Ledger Management page (#264)
nanocurrency/nano-docs: docs/running-a-node/ledger-management.md

# Ledger Management

The node automatically manages the full Nano ledger in the `data.ldb` file which can be found in the data folder at these locations:

--8<-- "folder-locations.md"

This file will grow in size as the ledger does. As of April 2020 there are over 49 million blocks in the ledger which requires at least 26GB of free space. See [hardware recommendations](/running-a-node/node-setup/#hardware-recommendations) for more preferred node specs.

!!! warning "RocksDB uses many files"

> 	The above details are for the default LMDB database setup. If using RocksDB, please note that it uses potentially 100s of SST files to manage the ledger so details should followed from the [RocksDB Ledger Backend](#rocksDB-ledger-backend) section below.

Reviewer: details should **be** followed

Author: Done

## Bootstrapping

> When starting a new node the ledger must be downloaded and kept updated in order to participate on the network properly. This is done automatically via bootstrapping - the node downloads and verifies blocks from other nodes across the network. This process can take hours to days to complete depending on network conditions and [hardware specifications](/running-a-node/node-setup/#hardware-recommendations).

Reviewer: Would add here that the operator should avoid restarting the node unless it's very clear from calls to RPC block_count that both block counts are stuck for a long time. It simply delays the bootstrap further.

Author: Callout added

### Tuning options

> **TODO**: Consider adding notes about tuning here and link to related config entries that need to be updated on configuration page: `bootstrap_connections_max` and `bootstrap_initiator_threads`? Maybe: "Depending on machine and networking resources, the bootstrap performance can be improved by updating X and X configuration values. The additional resource usage these options cause should be considered, especially if left during normal operation (after initial bootstrap is complete)."

Reviewer: I think this is good to have here, and specifically those two values you mentioned. My advice would be to not go beyond 64 `bootstrap_connections_max` and 4 `bootstrap_initiator_threads`, even for higher end nodes. That should be already plenty fast!

We also have the `--fast_bootstrap` CLI but think we should not mention here, due to causing unconfirmed blocks to not be deleted and increasing the ledger size.

Author: Done

## Ledger Fast Sync

!!! tip "Always backup your ledgers file"
	Whenever you are attempting to change the ledger, it is highly recommended you create backups of the existing `data.ldb` file to ensure you have a rollback point if issues are encountered.

To avoid bootstrapping times, a ledger file (`data.ldb`) can be downloaded off-chain and added to the data file used by the node. The Nano Foundation provides a daily ledger file download in the #ledger channel of our [Discord server](https://chat.nano.org). This is posted by `SergSW` and contains checksums for validation.

Before using this method there are a few considerations to ensure it is done safely:

### Data source
Make sure you trust the source providing the data to you. If you are unfamiliar with the individual or organization providing the ledger, consider other options for the data or fallback to the default of [bootstrapping](#bootstrapping) from the network.

> ### Voting weights

Reviewer: Could become more clear with **Representative voting weights**

Author: Done

Blocks are confirmed using the voting weight of representatives and these weights are determined by the account balances assigned to those representatives. In addition, the node releases contain a hard-coded set of representative weights captured at the time of release.

If a new ledger is downloaded it is recommended to compare the voting weights it contains against the hard-coded values in the node. To do this, the [--compare_rep_weights](/commands/command-line-interface/#-compare_rep_weights) CLI (_available in V21+ only_) can be run to get an output of the variance.

> Although the variance threshold considered to be safe will depend on freshness of the ledger file and network weight variations since the node release date, in general more than a 10% difference should be scrutinized further. If you need support in evaluating join the [Node and Representative Management category](https://forum.nano.org/c/node-and-rep/8) on the [Nano Forums](https://forum.nano.org).

Reviewer: The variance isn't available in that command currently, however it provides the total mismatched weight and the mean. https://github.com/nanocurrency/nano-node/pull/2719. Not sure how what thresholds should be considered though.

Reviewer: Standard deviation is now part of the output here.

Author: I am not certain of the right thresholds here either. Considering output such as:

```
ledger weight 115,490,317 Mnano
hardcoded weight 122,578,472 Mnano
mismatched
        total 26,997,517 Mnano
        samples 141
        mean 191,471 Mnano
        sigma 729,315 Mnano
```

can you breakdown what each of these pieces means so we can sort an explanation of how to evaluate?

Reviewer: Sure, the difference between total hardcoded and total ledger weight means that more weight is now undelegated (due to pending weight, right?) than in the hardcoded weights. A very large difference here (let's say 20M) and I would be skeptical.

Mismatched is based on individual samples: for every hardcoded rep weight, is it on the provided ledger? Each sample is the absolute difference between those. Total is the accumulation of those differences. As money moves around it is definitely expected to see fluctuations here, but in normal conditions - if weights are well distributed - they shouldn't be large (would indicate one representative receiving a lot of weight).

Sigma is the standard deviation if the sampled mismatched weight is modeled as a gaussian distribution - which it isn't, because it allows negative weight, but good enough for the purpose. [This page](https://en.wikipedia.org/wiki/68%E2%80%9395%E2%80%9399.7_rule) has some useful explanations.

Check the page above. Do you think printing out explicitly any values above 3-sigma (by definition should not be many, 1 or 2 max) is useful here? It could indicate forgery if there is one specific representative that gained a very significant amount of weight. These could then be checked on third party websites.

Moreover, would it be useful to have every mismatch sample printed out, perhaps in the log instead?

Author: 1. Scan the ledger for integrity using https://docs.nano.org/commands/command-line-interface/#-debug_validate_blocks - if issues are highlighted here inspect carefully and consider looking for alternative sources of a ledger as issues here may be a strong indication of tampering with the ledger.
1. Run `--compare_rep_weights` and review the "outliers" and "new reps" with large amounts, using public explorers to inspect the representatives to make sure the weight changes identified for them match up and that they appear to be valid, participating representatives.

Caveat that this tool isn't giving an absolute answer about issues with a ledger, but can help to identify when additional investigation could be needed. Downloaded ledgers should only ever be used by entities you trust and know as honest actors on the network.

Author: I will make adjustments and circulate changes here one latest changes are made per our discussion today.

Author: @guilhermelawless Adjustments made based on latest changes to https://github.com/nanocurrency/nano-node/pull/2719, kept it pretty general and thinking we would likely add some details in the CLI command area about the output. Thoughts?

Reviewer: Added one comment directly in the changes but otherwise LGTM. Was already thinking of further details in the CLI itself 👍

!!! tip "Updating the node may require a lengthy ledger upgrade"
	Read the [guide](#updating-the-node) further down this page for some tips on how to minimize downtime during an update.

---

## Bootstrapping

When starting a new node the ledger must be downloaded and kept updated in order to participate on the network properly. This is done automatically via bootstrapping - the node downloads and verifies blocks from other nodes across the network. This process can take hours to days to complete depending on network conditions and [hardware specifications](/running-a-node/node-setup/#hardware-recommendations).

!!! warning "Restarting node during bootstrapping not recommended"
	It is **highly recommended to avoid restarting the node during bootstrapping** as this can cause extra delays in the syncing process. An exception can be made when it is very clear from calls to the [`block_count`](/commands/rpc-protocol/#block_count) RPC that block counts are stuck for multiple hours.

### Tuning options

Depending on machine and networking resources, the bootstrap performance can be improved by updating the following [configuration](/running-a-node/configuration/) values:

> * `node.bootstrap_connections_max`: up to max of `64`
> * `node.bootstrap_initiator_threads`: up to max of `4`

Reviewer: ```suggestion
* `node.bootstrap_connections_max`: up to max of `128`
* `node.bootstrap_connections`: up to max of `16`
```

# Ledger Management

## Ledger file

The node automatically manages the full Nano ledger in the `data.ldb` file which can be found in the data folder at these locations:

--8<-- "folder-locations.md"

This file will grow in size as the ledger does. As of April 2020 there are over 49 million blocks in the ledger which requires at least 26GB of free space. See [hardware recommendations](/running-a-node/node-setup/#hardware-recommendations) for more preferred node specs.

!!! warning "RocksDB uses many files"

> 	The above details are for the default LMDB database setup. If using RocksDB, please note that it uses potentially 100s of SST files to manage the ledger so details should be followed from the [RocksDB Ledger Backend](#rocksDB-ledger-backend) section below.

Reviewer: ```suggestion
	The above details are for the default LMDB database setup. If using RocksDB, please note that it uses potentially 100s of SST files to manage the ledger so details should be followed from the [RocksDB Ledger Backend](#rocksdb-ledger-backend) section below.
```

### Validating blocks and voting weights
Blocks are confirmed using the voting weight of representatives and these weights are determined by the account balances assigned to those representatives. In addition, the node releases contain a hard-coded set of representative weights captured at the time of the node release to help this process during bootstrapping.

If looking to use a downloaded ledger there is a risk of it providing inaccurate representative voting weights. Although the potential impacts of this are minimal, below are some recommended steps to take which can help provide additional confidence the ledger can be used.

1. **Scan the ledger for integrity using the [`--debug_validate_blocks`](/commands/command-line-interface/#-debug_validate_blocks) CLI command**. If issues are found they should be inspected carefully and alternative sources of a ledger may need to be considered as failures with this command have a high chance of indicating potentially malicious behavior.

> 1. **Review the differences in representative voting weights by running the [`--compare_rep_weights`](/commands/command-line-interface/#-compare_rep_weights) CLI command** with the new ledger in the default data folder (old ledger backed up) or in a different data folder by using the optional `--data_path` argument. This will compare the new ledger voting weights against the hardcoded values in the node (set at the time of release). See the [CLI command]((/commands/command-line-interface/#-compare_rep_weights)) for details on the output with special attention paid to entries in the `outliers` and `newcomers` sections. By inspecting those addresses in public explorers such as [Nanocrawler.cc](https://nanocrawler.cc), this can help to determine if voting weight may have been manipulated in the downloaded ledger.

Reviewer: ```suggestion
1. **Review the differences in representative voting weights by running the [`--compare_rep_weights`](/commands/command-line-interface/#-compare_rep_weights) CLI command** with the new ledger in the default data folder (old ledger backed up) or in a different data folder by using the optional `--data_path` argument. This will compare the new ledger voting weights against the hardcoded values in the node (set at the time of release). See the [CLI command](/commands/command-line-interface/#-compare_rep_weights) for details on the output with special attention paid to entries in the `outliers` and `newcomers` sections. By inspecting those addresses in public explorers such as [Nanocrawler.cc](https://nanocrawler.cc), this can help to determine if voting weight may have been manipulated in the downloaded ledger.
```

---

## Updating the node

Occasionally, updating to the [latest node version](/releases/node-releases/#current-release) requires upgrading the existing ledger which can have the following effects:

- Significant downtime, from a few minutes to several hours, during which the node RPC is not accessible and no voting occurs. The upgrade is especially slower if the ledger is not on an SSD.
- Temporary increased disk space usage - up to 3x the current ledger size in total (e.g. 60GB for a 20GB ledger)

In order to minimize downtime, consider performing the update in a different machine, and replacing the [ledger file](#ledger-file) once complete. Note the following instructions, where **Machine A** has the node and ledger, and **Machine B** will be updating it.

1. Create a directory `/home/<user>/Nano_Update` on Machine B.
1. Stop the node on Machine A.
1. Copy the `data.ldb` [file](#ledger-management) from Machine A to `/home/<user>/Nano_Update/data.ldb` on Machine B.
1. Start the node again on Machine A.

> 1. Download the [latest node version](https://github.com/nanocurrency/nano-node/releases/latest). For the purposes of this guide, using a binary is easier.

Reviewer: ```suggestion
1. Download the [latest node version](/releases/node-releases/#current-release). For the purposes of this guide, using a binary is easier.
```

I wasn't sure when I wrote this guide, but do we want to link to the docs instead?

> 1. Launch the node as follows (varies based on your operating system): `./nano_node --daemon --data_path /home/<user>/Nano_Update --config node.logging.log_to_cerr=true`
> 1. The message *"Upgrade in progress..."* will be displayed if a ledger upgrade is required.
> 1. The upgrade is finished when new messages start appearing on the screen. At that point, press Ctrl+C to stop the node.

Reviewer: ```suggestion
1. Run the following command (varies based on your operating system): `./nano_node --debug_block_count --data_path /home/<user>/Nano_Update --config node.logging.log_to_cerr=true`
1. The message *"Upgrade in progress..."* will be displayed if a ledger upgrade is required. Wait until the command finishes and **do not stop the upgrade preemptively**.
```

1. Copy `/home/<user>/Nano_Update/data.ldb` from Machine B to a temporary location on Machine A. **do not overwrite data.ldb on Machine A while the node is running**.
1. Stop and **upgrade** to the latest node version on Machine A as you would do normally.
1. Stop the node on Machine A in case upgrading restarted it.
1. Replace `/home/<user>/Nano/data.ldb` with the transferred file.
1. Restart the node.

> In the event that you are unable to upgrade the ledger on another machine but would still like to minimize downtime, consider [obtaining the ledger from another source](#ledger-fast-sync) as a last resource.

Reviewer: ```suggestion
In the event that you are unable to upgrade the ledger on another machine but would still like to minimize downtime, consider [obtaining the ledger from another source](#downloaded-ledger-files) as a last resource.
```

---

## RocksDB Ledger Backend

!!! warning "RocksDB is experimental, do not use in production"
	RocksDB is being included in _V20.0_ as experimental only. Future versions of the node may allow for production use of RocksDB, however old experimental RocksDB ledgers are not guarenteed to be compatible and may require resyncing from scratch.

	If you are testing RocksDB and want to discuss results, configurations, etc. please join the forum topic here: https://forum.nano.org/t/rocksdb-ledger-backend-testing/111

The node ledger currently uses LMDB (Lightning memory-mapped database) by default as the data store. As of _v20+_ the option to use RocksDB becomes available as an experimental option.
This document will not go into much detail about theses key-value data stores as there is a lot of information available online.
It is anticipated that bootstrapping will be slower using RocksDB during the initial version at least, but live traffic should be faster due to singluar writes being cached in memory and flushed to disk in bulk.

> Using RocksDB requires a few extra steps as it is an externally required dependency which requires a recent version of RocksDB, so older repositories may not be sufficient, it also requires `zlib`. If using the docker node, can skip to [Enable RocksDB](#enable-rocksdb):  

Reviewer: ```suggestion
Using RocksDB requires a few extra steps as it is an externally required dependency which requires a recent version of RocksDB, so older repositories may not be sufficient, it also requires `zlib`. If using the docker node, skip to [Enable RocksDB](#enable-rocksdb).
```

The additional resource usage these options cause should be considered, especially if left during normal operation (after initial bootstrap is complete).

---

## Downloaded ledger files

!!! tip "Always backup your ledgers file"
	Whenever you are attempting to change the ledger, it is highly recommended you create backups of the existing `data.ldb` file to ensure you have a rollback point if issues are encountered.

To avoid bootstrapping times, a ledger file (`data.ldb`) can be downloaded off-network and added to the data file used by the node. This process is sometimes referred to as a "fast sync". The Nano Foundation provides a daily ledger file download in the #ledger channel of our [Discord server](https://chat.nano.org). This is posted by `SergSW` and contains checksums for validation.

Before using this method there are a few considerations to ensure it is done safely:

### Data source
Make sure you trust the source providing the data to you. If you are unfamiliar with the individual or organization providing the ledger, consider other options for the data or fallback to the default of [bootstrapping](#bootstrapping) from the network.

> ### Representative voting weights

Reviewer: ```suggestion
### Validating blocks and voting weights
```

Blocks are confirmed using the voting weight of representatives and these weights are determined by the account balances assigned to those representatives. In addition, the node releases contain a hard-coded set of representative weights captured at the time of the node release to help this process during bootstrapping.

If looking to use a downloaded ledger there is a risk of it providing inaccurate representative voting weights. Although the potential impacts of this are minimal, below are some recommended steps to take which can help provide additional confidence the ledger can be used.

1. **Scan the ledger for integrity using the [`--debug_validate_blocks`](/commands/command-line-interface/#-debug_validate_blocks) CLI command**. If issues are found they should be inspected carefully and alternative sources of a ledger may need to be considered as failures with this command have a high chance of indicating potentially malicious behavior.

> 1. With the new ledger in the data folder (don't forget to backup the previous ledger!), **review the differences in representative voting weights by running the [`--compare_rep_weights`](/commands/command-line-interface/#-compare_rep_weights) CLI command**. This will compare the new ledger voting weights against the hardcoded values in the node (set at the time of release). See the CLI command for details on the output, but special attention should be paid to entries in the `outliers` section, which indicates large differences between the hard coded and new ledger rep weights, and the `newcomers` section, which highlights large voting weights showing up for new representatives. By inspecting the output addresses in public explorers such as [Nanocrawler.cc](https://nanocrawler.cc), this can help to determine if voting weight may have been manipulated in the downloaded ledger.

Reviewer: `--data_path` can be used to point the command to another ledger so that there's no need to backup the current one and move things around. For a ledger in `/path/to/ledger/data.ldb` , the ccommand would be `--compare_rep_weights --data_path /path/to/ledger/`

Author: Tweaked this to mentioned `--data_path` and overall simplify things to push more on the CLI description.
